### PR TITLE
Avoid serializer multi-threaded cache issues

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/corefx/issues/36547

The previous code has a List<> containing cached properties which was initialized on the first deserialize. If there are many threads trying to deserialize the same POCO type (during this warm-up phase) it is possible that the List<> could get corrupted or throw an exception during an Add() or Item().

The fix is to use a stack-based List<> cache (during the warm-up phase) and assign that to the POCO type cache only when it is completed, overwriting any other thread's cache if there happens to be a race condition.

### CODE REVIEW

File: src/System.Text.Json/tests/Serialization/ThreadingTests.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                testObj.Initialize();
+                testObj.Verify();
+
+                string json = JsonSerializer.ToString(testObj);
+                SimpleTestClass testObjDeserialized = JsonSerializer.Parse<SimpleTestClass>(json);
+                testObjDeserialized.Verify();
+            };
+
+            Task[] tasks = new Task[8];
+            for (int i = 0; i <tasks.Length; i++)
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/ThreadingTests.cs]: nit: spacing

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 2]
```diff
@@ -15,8 +15,10 @@ internal sealed partial class JsonClassInfo
         // The length of the property name embedded in the key (in bytes).
         private const int PropertyNameKeyLength = 6;
 
-        private readonly List<PropertyRef> _property_refs = new List<PropertyRef>();
-        private readonly List<PropertyRef> _property_refs_sorted = new List<PropertyRef>();
+        private readonly List<PropertyRef> _propertyRefs = new List<PropertyRef>();
+
+        // Cache of properties by first ordering. Use an array for highest performance.
+        private volatile PropertyRef[] _propertyRefsSorted = null;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Why does this need to be `volatile`?

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 3]
```diff
@@ -28,6 +30,15 @@ internal sealed partial class JsonClassInfo
 
         public Type Type { get; private set; }
 
+        internal void UpdateSortedPropertyCache(ref ReadStackFrame frame)
+        {
+            // Set the sorted property cache. Overwrite any existing cache which can occur in multi-threaded cases.
+            if (frame.PropertyRefCache != null)
+            {
+                _propertyRefsSorted = frame.PropertyRefCache.ToArray();
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Do we need some kind of synchronization/locking mechanism here?

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 4]
```diff
@@ -66,85 +77,104 @@ internal JsonClassInfo(Type type, JsonSerializerOptions options)
...
-                if (TryIsPropertyRefEqual(_property_refs, propertyName, key, i, out info))
+                PropertyRef propertyRef = _propertyRefs[i];
+                if (TryIsPropertyRefEqual(ref propertyRef, propertyName, key, out info))
                 {
                     break;
                 }
             }
 
-            if (info != null)
+            if (!hasPropertyCache)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Can this be an else condition to `if (hasPropertyCache)` on line 90, before the loop?

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 3 above)
Comment (stephentoub) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: > Do we need some kind of synchronization/locking mechanism here?

That's why the above is volatile.  It avoids out-of-order issues with regards to the data written to the array being published prior to the reference to the array, and similar for the read side.  Technically the volatile isn't necessary due to additional constraints coreclr puts in place, but it is necessary according to the ECMA CLI spec.  And as long as the results here are idempotent, it doesn't matter whether multiple threads race to set the field.

**Comment by steveharter** [comment]
I will be adding a new commit to deal with a large number of properties to any performance issues (since the code overlaps with changes here) - i.e. change from list to dictionary at a certain threshold.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 4 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: No. I added a comment to explain why (race conditions).

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs (hunk 2 above)
Comment (steveharter) [reply src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: Multiple threads may attempt to overwrite the same field, so this makes sure the value is set atomically across word boundaries.

**Comment by wfurt** [comment]
seems like the test is not stable:
https://mc.dot.net/#/product/netcore/30/source/official~2Fdotnet~2Fcorefx~2Frefs~2Fheads~2Fmaster/type/test~2Ffunctional~2Fcli~2F/build/20190404.4/workItem/System.Text.Json.Tests

passing only with retires on some platforms and failed on 2 others.

**Pull request merged**